src/wallet/keyman.cpp

```diff
@@ -0,0 +1,278 @@
+    if (!m_map_keys.empty()) {
+        return false;
+    }
+
+    m_map_crypted_keys[key_id] = make_pair(pubkey, ckey);
+    return true;
+}
+
+void KeyManager::LoadHDKey(const CKeyID& key_id, const CExtPubKey& xpub)
+{
+    LOCK(cs_keyman);
+    m_map_xpubs[key_id] = xpub;
+}
+
+bool KeyManager::CheckDecryptionKey(const CKeyingMaterial& master_key, bool accept_no_keys)
+{
+    LOCK(cs_keyman);
+    if (!m_map_keys.empty()) {
+        return false;
+    }
+
+    bool keyPass = m_map_crypted_keys.empty(); // Always pass when there are no encrypted keys
+    bool keyFail = false;
+    for (const auto& mi : m_map_crypted_keys) {
+        const CPubKey& pubkey = mi.second.first;
+        const std::vector<unsigned char>& crypted_secret = mi.second.second;
+        CKey key;
+        if (!DecryptKey(master_key, crypted_secret, pubkey, key)) {
+            keyFail = true;
+            break;
+        }
+        keyPass = true;
+        if (m_decryption_thoroughly_checked) {
+            break;
+        }
+    }
+    if (keyPass && keyFail) {
+        LogPrintf("The wallet is probably corrupted: Some keys decrypt but not all.\n");
+        throw std::runtime_error("Error unlocking wallet: some keys decrypt but not all. Your wallet file may be corrupt.");
+    }
+    if (keyFail || (!keyPass && !accept_no_keys)) {
+        return false;
+    }
+    m_decryption_thoroughly_checked = true;
+    return true;
+}
+
+bool KeyManager::Encrypt(const CKeyingMaterial& master_key, WalletBatch* batch)
+{
+    LOCK(cs_keyman);
+
+    // Nothing to encrypt
+    if (m_map_keys.empty()) {
+        return true;
+    }
+
+    if (!m_map_crypted_keys.empty()) {
+        return false;
+    }
+
+    for (const auto& [id, key] : m_map_keys)
+    {
+        CPubKey pubkey = key.GetPubKey();
+        CKeyingMaterial secret(key.begin(), key.end());
+        std::vector<unsigned char> crypted_secret;
+        if (!EncryptSecret(master_key, secret, pubkey.GetHash(), crypted_secret)) {
+            return false;
+        }
+        m_map_crypted_keys[id] = make_pair(pubkey, crypted_secret);
+        batch->WriteCryptedKeyManKey(pubkey, crypted_secret);
+    }
+    m_map_keys.clear();
+    return true;
+}
+
+std::map<CKeyID, CKey> KeyManager::GetKeys() const
+{
+    AssertLockHeld(cs_keyman);
+    if (m_storage.HasEncryptionKeys() && !m_storage.IsLocked()) {
+        std::map<CKeyID, CKey> keys;
+        for (const auto& [id, key_pair] : m_map_crypted_keys) {
+            const auto& [pubkey, crypted_secret] = key_pair;
+            CKey key;
+            bool ok = DecryptKey(m_storage.GetEncryptionKey(), crypted_secret, pubkey, key);
+            assert(ok);
+            keys[id] = key;
+        }
+        return keys;
+    }
+    // If the wallet is encrypted and locked, then this will just be an empty map
+    return m_map_keys;
+}
+
+bool KeyManager::HavePrivateKeys() const
+{
+    LOCK(cs_keyman);
+    return m_map_keys.empty() || m_map_crypted_keys.empty();
```

Comment:
i think you meant 
```suggestion
    return !m_map_keys.empty() || !m_map_crypted_keys.empty();
```

Reply:
also, maybe `HasPrivateKeys` is better than `HavePrivateKeys` ?

Reply:
Oops, yes.

> also, maybe `HasPrivateKeys` is better than `HavePrivateKeys` ?

I prefer to keep it the same for consistent naming.

Reply:
Yikes. Can you add a test for this?